wit/wasi-keyvalue.wit

```diff
@@ -0,0 +1,157 @@
+// A generic keyvalue interface for WASI.
+interface types {
+	/// A bucket is a collection of key-value pairs. Each key-value pair is stored
+	/// as a entry in the bucket, and the bucket itself acts as a collection of all
+	/// these entries. 
+	///
+	/// It is worth noting that the exact terminology for bucket in key-value stores
+	/// can very depending on the specific implementation. For example,
+	/// 1. Amazon DynamoDB calls a collection of key-value pairs a table
+	/// 2. Redis has hashes, sets, and sorted sets as different types of collections
+	/// 3. Cassandra calls a collection of key-value pairs a column family
+	/// 4. MongoDB calls a collection of key-value pairs a collection
+	/// 5. Riak calls a collection of key-value pairs a bucket
+	/// 6. Memcached calls a collection of key-value pairs a slab
+	/// 7. Azure Cosmos DB calls a collection of key-value pairs a container
+	///
+	/// In this interface, we use the term `bucket` to refer to a collection of key-value
+	// Soon: switch to `resource bucket { ... }`
+	type bucket = u32
+	drop-bucket: func(bucket: bucket)
+	open-bucket: func(name: string) -> result<bucket, error>
+
+	/// A key is a unique identifier for a value in a bucket. The key is used to
+	/// retrieve the value from the bucket.
+	type key = string
+
+	/// A list of keys
+	type keys = list<key>
+	
+	/// An error resource type for keyvalue operations.
+	/// Currently, this provides only one function to return a string representation
+	/// of the error. In the future, this will be extended to provide more information
+	/// about the error.
+	// Soon: switch to `resource error { ... }`
+	type error = u32
+	drop-error: func(error: error)
+	trace: func(error: error) -> string
```

Comment:
I think this should be brought out of here and put into a wasi-error proposal~ I suggested this at one of the last WASI meetings of last year, and I think it would be good to act on it, 'cause all of our proposals will use this (will try to bring it up in the next one).